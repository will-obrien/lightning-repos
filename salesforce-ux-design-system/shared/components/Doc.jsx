// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

import cx from 'classnames';
import get from 'lodash.get';
import React from 'react';

import '../styles/doc.scss';
import { flattenElement, mapElement } from '../utils/react';

import Example from './Example';

export const createAnchor = (type, id) =>
  React.createElement(
    'a',
    {
      'aria-label': 'Jump to section',
      className: 'doc doc-anchor',
      'data-level': type[1],
      href: `#${id}`
    },
    '#'
  );

export const tags = [
  'p',
  'div',
  'a',
  'em',
  'strong',
  'ol',
  'ul',
  'li',
  'code',
  'blockquote',
  'pre',
  'tr',
  'td',
  'th',
  'table',
  'thead',
  'tbody',
  'h1',
  'h2',
  'h3',
  'h4',
  'h5',
  'h6'
];

export const factories = tags.reduce(
  (factories, type) =>
    Object.assign({}, factories, {
      [type]: (props, ...children) => {
        if (/h[1-6]/.test(type)) {
          children.push(createAnchor(type, props.id));
        }
        return React.createElement(
          type,
          {
            ...props,
            className: cx('doc', props.className)
          },
          ...children
        );
      }
    }),
  {}
);

export const createTableOfContents = (() => {
  const createNode = (level, element) => {
    const { children, id } = element.props;
    return {
      children: [],
      id,
      level,
      title:
        typeof children === 'string'
          ? children
          : Array.isArray(children) ? children[0] : children
    };
  };
  return element => {
    let keyPath = [];
    let tree = {
      level: 0,
      children: []
    };
    flattenElement(mapElement(element, e => (e.type === Example ? null : e)))
      .filter(e => /^h[1-6]/.test(e.type) && e.props.id)
      .forEach(element => {
        let level = parseInt(element.type[1], 10);
        let levelLast = get(tree, keyPath.concat(['level']));
        if (level <= levelLast) {
          while (keyPath.length) {
            const heading = get(tree, keyPath);
            keyPath = keyPath.slice(0, -2);
            if (heading.level === level) break;
          }
        }
        const children = get(tree, keyPath.concat(['children']));
        children.push(createNode(level, element));
        keyPath.push('children', children.length - 1);
      });
    return tree;
  };
})();

export default class Doc extends React.Component {
  render() {
    return <div {...this.props} />;
  }
}
