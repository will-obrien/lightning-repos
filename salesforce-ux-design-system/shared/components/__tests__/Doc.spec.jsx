// Copyright (c) 2015-present, salesforce.com, inc. All rights reserved
// Licensed under BSD 3-Clause - see LICENSE.txt or git.io/sfdc-license

/* eslint-env jest */

import React from 'react';

import { createTableOfContents } from '../Doc';

const mdx = (
  <div>
    <h1 id="a">A</h1>
    <p>Lorem ipsum dolor sit amemt</p>
    <h2 id="aa">AA</h2>
    <p>Lorem ipsum dolor sit amemt</p>
    <h1 id="b">B</h1>
    <p>Lorem ipsum dolor sit amemt</p>
    <h2 id="bb">BB</h2>
    <p>Lorem ipsum dolor sit amemt</p>
    <h3 id="bbb">BBB</h3>
    <p>Lorem ipsum dolor sit amemt</p>
  </div>
);

it('createTableOfContents', () => {
  expect(createTableOfContents(mdx)).toMatchSnapshot();
});
