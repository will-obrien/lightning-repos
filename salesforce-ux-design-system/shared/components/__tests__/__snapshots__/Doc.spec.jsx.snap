// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createTableOfContents 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [],
          "id": "aa",
          "level": 2,
          "title": "AA",
        },
      ],
      "id": "a",
      "level": 1,
      "title": "A",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "id": "bbb",
              "level": 3,
              "title": "BBB",
            },
          ],
          "id": "bb",
          "level": 2,
          "title": "BB",
        },
      ],
      "id": "b",
      "level": 1,
      "title": "B",
    },
  ],
  "level": 0,
}
`;
